Objects/dictobject.c

```diff
@@ -1285,6 +1285,20 @@ ensure_shared_on_resize(PyDictObject *mp)
 #endif
 }
 
+static inline void
+ensure_shared_on_keys_version_assignment(PyDictObject *mp)
+{
+    ASSERT_DICT_LOCKED((PyObject *) mp);
+    #ifdef Py_GIL_DISABLED
+    if (!_Py_IsOwnedByCurrentThread((PyObject *)mp) && !IS_DICT_SHARED(mp)) {
```

Comment:
I think this check should just be: `if (!IS_DICT_SHARED(mp))`. 

The `_Py_IsOwnedByCurrentThread()` check here might be okay with some other changes and a few assumptions (like specialization remains per-thread), but I think it's easier to reason about if this is just marks the dict as shared if it's not already.

Reply:
🤦 yep